pytensor/xtensor/type.py

```diff
@@ -38,17 +42,329 @@
+        return DIM_LENGTH_TYPE.filter(
+            data, strict=strict, allow_downcast=allow_downcast
+        )
+
+    def filer_variable(self, other, allow_convert=True):
+        """Filter a variable to ensure it is a DimVariable."""
+        if not isinstance(other, Variable):
+            raise ValueError()
+
+        if isinstance(other.type, DimType):
+            return other
+
+        if allow_convert:
+            other2 = self.convert_variable(other)
+            if other2 is not None:
+                return other2
+
+        raise TypeError(
+            f"Cannot convert Type {other.type} (of Variable {other}) into Type {self}."
+        )
+
+    def __repr__(self) -> str:
+        props = []
+        for prop in self.__props__:
+            if not hasattr(self, prop):
+                raise AttributeError(
+                    f"{self.__class__.__name__} has no property '{prop}' even though it is listed in __props__"
+                )
+            value = getattr(self, prop)
+            if value is None:
+                continue
+            if prop == "name":
+                props.insert(0, f"{value}")
+            elif prop == "uuid":
+                props.append("uuid=?")
+            else:
+                props.append(f"{prop}={value!r}")
+        return f"{self.__class__.__name__}({', '.join(props)})"
+
+    def dim_compatible(self, other: DimType):
+        """Test if the dimension is compatible with other dimensions.
+
+        If two dimensions are compatible, they must have a common
+        dimension that they can broadcast to. Tensors can not contain
+        any dimensions that are compatible.
+
+        dim compatibility *must* me reflexive, symmetric and transitive.
+
+        It defaults to dim equality, but can be overridden by subclasses.
+        """
+        return self == other
+
+    def broadcasted_dim_type(self, *other: DimType) -> DimType | None:
+        """Find the smallest dimension that all dimensions can broadcast to.
+
+        Note, that this does not correspond to the usual numpy broadcasting,
+        but will be used mostly to broadcast dimensions that are subsets
+        of some larger dimension.
+
+        If the dimensions are not compatible, it returns None.
+        """
+        if all(self.dim_compatible(o) for o in other):
+            return self
+        return None
+
+    def broadcast_dim(self, dim_var: DimVariable, target_type: DimType) -> DimVariable:
+        """Broadcast this dimension to the given broadcast_dim.
+
+        If the dimensions are not compatible, it raises a ValueError.
+        """
+        if target_type == self:
+            return dim_var
+
+        if not self.dim_compatible(target_type):
+            raise ValueError(
+                f"Cannot broadcast {self} to {target_type}. "
+                "Dimensions must be compatible."
+            )
+        raise NotImplementedError("Subclass did not implent dim broadcasting")
+
+
+class BasicDim(DimType):
```

Comment:
Nit: I would perhaps split dim_type / var_type into separate files, this one is already pretty long as is